ndb/src/google/cloud/ndb/model.py

```diff
@@ -3676,6 +3676,80 @@ class a different name when stored in Google Cloud Datastore than the
         """
         return cls.__name__
 
+    def __hash__(self):
+        """Not implemented hash function.
+
+        Raises:
+            TypeError: Always, to emphasize that entities are mutable.
+        """
+        raise TypeError("Model is not immutable")
+
+    def __eq__(self, other):
+        """Compare two entities of the same class for equality."""
+        if type(other) is not type(self):
+            return NotImplemented
+
+        if self._key != other._key:
+            return False
+
+        return self._equivalent(other)
+
+    def _equivalent(self, other):
+        """Compare two entities of the same class, excluding keys.
+
+        Args:
+            other (Model): An entity of the same class. It is assumed that
+                the type and the key of ``other`` match the current entity's
+                type and key (and the caller is responsible for checking).
+
+        Returns:
+            bool: Indicating if the current entity and ``other`` are
+            equivalent.
+
+        Raises:
+            NotImplementedError: If the type's don't match.
+        """
+        if type(other) is not type(self):
+            raise NotImplementedError(
+                "Cannot compare different model classes. {} is "
+                "not {}".format(type(self).__name__, type(other).__name__)
+            )
+
+        if set(self._projection) != set(other._projection):
+            return False
+
+        prop_names = set(self._properties.keys())
+        # Restrict properties to the projection if set.
+        if self._projection:
+            prop_names = set(self._projection)
+
+        for name in prop_names:
+            value = self._properties[name]._get_value(self)
+            if value != other._properties[name]._get_value(other):
+                return False
+
+        return True
+
+    def __ne__(self, other):
```

Comment:
I *think* this isn't strictly necessary. Python can negate __eq__ without an explicit __neq__.

~~~~
Python 3.7.0 (default, Oct 15 2018, 12:13:42) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...   def __eq__(self, other):
...       return other == 5
... 
>>> a = A()
>>> a == 5
True
>>> a == 6
False
>>> a != 6
True
>>> a != 5
False
>>> 
~~~~

Reply:
From https://docs.python.org/3/reference/datamodel.html:

> By default, `__ne__()` delegates to `__eq__()` and inverts the result unless it is `NotImplemented`. There are no other implied relationships among the comparison operators, for example, the truth of `(x<y or x==y)` does not imply `x<=y`. To automatically generate ordering operations from a single root operation, see `functools.total_ordering()`.

Reply:
Interesting. I wasn't 100% sure, which is why I tried it in the REPL and pasted the result above. I didn't try in 2.7, because we don't need it.

Reply:
OK I removed all `__ne__` except for one that was doing something other than inverting `__eq__`